compiler/rustc_hir_typeck/src/op.rs

```diff
@@ -521,6 +521,63 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                    if op.node == hir::BinOpKind::Add {
+                        match (lhs_ty.kind(), rhs_ty) {
+                            (RawPtr(..), rhs_ty) if rhs_ty.is_integral() => {
+                                let lhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(lhs_expr.span);
+                                let rhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(rhs_expr.span);
+                                if let Ok(lhs_snip) = lhs_snip && let Ok(rhs_snip) = rhs_snip {
+                                    err.span_suggestion(
+                                        expr.span,
+                                        "consider using `wrapping_add` or `add` for pointer + {integer}",
+                                        format!("{lhs_snip}.wrapping_add({rhs_snip})"),
+                                        Applicability::MaybeIncorrect
+                                    );
+                                }
+                            }
+                            _ => {}
+                        }
+                    } else if op.node == hir::BinOpKind::Sub {
+                        match (lhs_ty.kind(), rhs_ty) {
+                            (RawPtr(..), rhs_ty) if rhs_ty.is_integral() => {
+                                let lhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(lhs_expr.span);
+                                let rhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(rhs_expr.span);
+                                if let Ok(lhs_snip) = lhs_snip && let Ok(rhs_snip) = rhs_snip {
+                                    err.span_suggestion(
+                                        expr.span,
+                                        "consider using `wrapping_sub` or `sub` for pointer - {integer}",
+                                        format!("{lhs_snip}.wrapping_sub({rhs_snip})"),
+                                        Applicability::MaybeIncorrect
+                                    );
+                                }
+                            }
+                            (RawPtr(..), rhs_ty) if rhs_ty.is_unsafe_ptr() => {
+                                let lhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(lhs_expr.span);
+                                let rhs_snip =
+                                    self.tcx.sess.source_map().span_to_snippet(rhs_expr.span);
+                                if let Ok(lhs_snip) = lhs_snip && let Ok(rhs_snip) = rhs_snip {
+                                    err.span_suggestion(
+                                        expr.span,
+                                        "consider using `offset_from` for pointer - pointer",
+                                        format!("unsafe {{ {lhs_snip}.offset_from({rhs_snip}) }}"),
```

Comment:
Is there a safe suggestion we could make here? I didn't find one immediately after poking around the docs a little...

Reply:
I couldn't find one either...

Reply:
you should at least add "if the pointers point to the same allocation"